packages/core/src/node/cluster/master-process.ts

```diff
@@ -8,13 +8,15 @@
 import { EventEmitter } from 'events';
 import { ServerWorker } from './server-worker';
 
-export type MasterProcessEvent = 'started' | 'restarted' | 'restarting';
+export type MasterProcessEvent = 'started' | 'restarted' | 'restarting' | 'exit';
 export class MasterProcess extends EventEmitter {
 
     protected serverWorker: ServerWorker | undefined;
 
     protected fork(): ServerWorker {
-        return new ServerWorker(() => this.restart());
+        const worker = new ServerWorker(() => this.restart());
+        worker.exit.then(() => this.emit('exit', worker));
```

Comment:
What about propagating the exit code? i.e.
`worker.exit.then(code => this.emit('exit', code))` and exiting the main process with the same code

Reply:
For `restart` this listener should be installed only a new worker is successfully initialized, i.e. at https://github.com/theia-ide/theia/pull/1893/files#diff-5ae26f56a8fb4b14d6f3f63382663531R54